### ISSUE TITLE
[BUG] Flytekit misuses unexpected fsspec lpath result

### ISSUE THREAD

**Issue by cgrass** [flyteorg/flyte#4700]
### Describe the bug

fsspec has unexpected behavior in some scenarios. calling get with a URL for `rpath` (remote path) and a filename for `lpath` (local path) results in `lpath` being used as a path dir. e.g., if `rpath == http://mydomain/output.pb` and `lpath == /var/output.pb` a file gets written at `/var/output.pb/output.pb`.

this causes an issue in the flytekit code because [remote._get_output_literal_map()](https://github.com/flyteorg/flytekit/blob/b5dbd857a5b81757d13d03116574ba705ea8798c/flytekit/remote/remote.py#L1900C1-L1901C1) creates `tmp_name` with the filename ("output.pb"), and then uses `tmp_name` as the location for the resulting file. in reality, that path is a dir that contains the actual file. trying to open `tmp_name` results in a `IsADirectoryError`.

a [discussion](https://flyte-org.slack.com/archives/CP2HDHKE1/p1698250094391179) in slack can add more context.

### Expected behavior

remote file gets written to `/var/output.pb`

### Additional context to reproduce

You can see the fsspec behavior using the unit test below
```
def test_get_output():
    kwargs = {}
    spec = fsspec.implementations.http.HTTPFileSystem(fsspec.filesystem("https", **kwargs))
    frompath = linkToUrlFileile
    topath = "/var/inputs10.txt"
    response = spec.get(frompath, topath, **kwargs)
    print(response)
```

e2e testing is much harder because it requires a specific flyte configuration (described in [this PR](https://github.com/flyteorg/flyte/pull/4629)). @wild-endeavor was able to reproduce using a local setup.

### Are you sure this issue hasn't been raised already?

- [X] Yes

### Have you read the Code of Conduct?

- [X] Yes

**Comment by wild-endeavor** [comment]
actually @cgrass could you check again?  i'm not seeing this behavior using plain fsspec

```
def test_get_output():
    from fsspec.implementations.http import HTTPFileSystem

    spec = HTTPFileSystem(fsspec.filesystem("https"))
    frompath = "http://www.ascii-art.de/ascii/s/startrek.txt"
    topath = "/Users/ytong/temp/2063.txt"
    response = spec.get(frompath, topath)
    print(response)
```

which version of fsspec do you have?  i'm on `2024.2.0`

**Comment by cgrass** [comment]
@wild-endeavor I see the same behavior in fsspec `2024.2.0`. your test above will create a dir at `/Users/ytong/temp/2063.txt` which will contain a file `startrek.txt`.

locally stored file at: `/Users/ytong/temp/2063.txt/startrek.txt`

**Comment by wild-endeavor** [comment]
@cgrass thanks for the clarification!  do you think your existing ticket on https://github.com/fsspec/filesystem_spec/issues/1505 covers this?  since this is a purely fsspec issue (like no flytekit is involved in the repro) it's a little hard to work around for us until it's fixed in fsspec.

**Comment by cgrass** [comment]
@wild-endeavor I agree the root problem is with fsspec behavior but there are a few options for flyte to harden the code and potentially get the implementation to work even with the current fsspec version.
1) try the list option (i see you already created a PR to test if that works)
2) update [remote._get_output_literal_map()](https://github.com/flyteorg/flytekit/blob/b5dbd857a5b81757d13d03116574ba705ea8798c/flytekit/remote/remote.py#L1900C1-L1901C1). I put together a simple POC when i first discovered the bug that was a little smarter about how `tmp_name` was created and passed around. that worked for my specific use case but needs to be tested for more general use.
3) use `get_file` instead of `get` when a single file is being fetched.
